File: module-app/src/main/java/maple/expectation/application/usecase/AuthPortAdapter.java
package maple.expectation.application.usecase;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import maple.expectation.core.port.inbound.AuthCommand;
import maple.expectation.core.port.inbound.AuthPort;
import maple.expectation.core.port.inbound.AuthResult;
import maple.expectation.core.port.inbound.TokenResult;
import maple.expectation.infrastructure.security.jwt.JwtTokenProvider;
import org.springframework.stereotype.Component;

/**
 * AuthPort 구현체 (ADR-005)
 *
 * <p>책임: 인증 관련 유스케이스 구현
 *
 * <p>V5 Migration (Issue #589): Redis 기반 세션/리프레시 토큰 저장소 제거 후 JWT-only 방식으로 단순화.
 *
 * <p>현재 상태: Refresh Token 기능 미지원 (Stateless JWT만 사용)
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class AuthPortAdapter implements AuthPort {

  private final JwtTokenProvider jwtTokenProvider;
  private static final String DEFAULT_ROLE = "USER";
  private static final long REFRESH_EXPIRES_IN = 604800L; // 7 days (not implemented yet)

  @Override
  public AuthResult login(AuthCommand command) {
    log.info("[AuthPort] Login attempt: userIgn={}", command.getUserIgn());

    // Generate session ID and fingerprint from apiKey
    String sessionId = generateSessionId(command.getApiKey());
    String fingerprint = generateFingerprint(command.getApiKey());

    // Determine role (check if admin)
    String role = determineRole(fingerprint);

    // Generate JWT token
    String accessToken = jwtTokenProvider.generateToken(sessionId, fingerprint, role);
    long expiresIn = jwtTokenProvider.getExpirationSeconds();

    log.info("[AuthPort] Login successful: sessionId={}, role={}", sessionId, role);

    return AuthResult.of(
        accessToken,
        expiresIn,
        role,
        fingerprint,
        "refresh-" + sessionId, // Placeholder refresh token
        REFRESH_EXPIRES_IN);
  }

  @Override
  public void logout(String sessionId) {
    log.info("[AuthPort] Logout: sessionId={}", sessionId);
    // Stateless JWT - no server-side session to invalidate
    // Token will expire naturally
  }

  @Override
  public TokenResult refresh(String refreshTokenId) {
    log.warn("[AuthPort] Token refresh not implemented yet. refreshTokenId={}", refreshTokenId);
    throw new UnsupportedOperationException(
        "Token refresh is not implemented in this version. Please login again.");
  }

  private String generateSessionId(String apiKey) {
    return "session-" + Math.abs(apiKey.hashCode()) + "-" + System.currentTimeMillis();
  }

  private String generateFingerprint(String apiKey) {
    // Simple fingerprint generation from API key
    return "fp-" + Integer.toHexString(apiKey.hashCode());
  }

  private String determineRole(String fingerprint) {
    // TODO: Check against admin allowlist
    return DEFAULT_ROLE;
  }
}

File: module-app/src/main/kotlin/maple/expectation/application/usecase/AuthPortAdapter.kt
package maple.expectation.application.usecase

import maple.expectation.core.port.inbound.AuthCommand
import maple.expectation.core.port.inbound.AuthPort
import maple.expectation.core.port.inbound.AuthResult
import maple.expectation.core.port.inbound.TokenResult
import maple.expectation.application.service.auth.ApiKeyValidator
import maple.expectation.infrastructure.security.jwt.JwtTokenProvider
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

/**
 * AuthPort 구현체 (ADR-005)
 *
 * 책임: 인증 관련 유스케이스 구현
 *
 * V5 Migration (Issue #589): Redis 기반 세션/리프레시 토큰 저장소 제거 후 JWT-only 방식으로 단순화.
 *
 * #667: Nexon API를 통한 실제 계정 검증. fingerprint = Nexon account_id.
 * 동일 계정의 다른 API Key라도 동일 account_id가 반환되어 identity가 보장됨.
 *
 * 현재 상태: Refresh Token 기능 미지원 (Stateless JWT만 사용)
 */
@Component
class AuthPortAdapter(
    private val jwtTokenProvider: JwtTokenProvider,
    private val apiKeyValidator: ApiKeyValidator,
) : AuthPort {

    override fun login(command: AuthCommand): AuthResult {
        log.info("[AuthPort] Login attempt: userIgn={}", command.userIgn)

        // #667: Nexon API 검증 → account_id + 소유 캐릭터 OCID 확보
        val validationResult = apiKeyValidator.validateAndVerifyOwnership(
            command.apiKey, command.userIgn,
        )
        val accountId = validationResult.accountId

        val sessionId = generateSessionId(accountId)
        val role = determineRole(accountId)

        // fingerprint = Nexon account_id (동일 계정 = 동일 ID, API Key 무관)
        val accessToken = jwtTokenProvider.generateToken(sessionId, accountId, role, command.userIgn)
        val expiresIn = jwtTokenProvider.getExpirationSeconds()

        log.info("[AuthPort] Login successful: sessionId={}, role={}, accountId={}", sessionId, role, accountId)

        return AuthResult.of(
            accessToken,
            expiresIn,
            role,
            accountId,
            "refresh-$sessionId",
            REFRESH_EXPIRES_IN,
        )
    }

    override fun logout(sessionId: String) {
        log.info("[AuthPort] Logout: sessionId={}", sessionId)
    }

    override fun refresh(refreshTokenId: String): TokenResult {
        log.warn("[AuthPort] Token refresh not implemented yet. refreshTokenId={}", refreshTokenId)
        throw UnsupportedOperationException(
            "Token refresh is not implemented in this version. Please login again.",
        )
    }

    private fun generateSessionId(accountId: String): String =
        "session-${kotlin.math.abs(accountId.hashCode())}-${System.currentTimeMillis()}"

    private fun determineRole(accountId: String): String {
        // TODO: Check against admin allowlist
        return DEFAULT_ROLE
    }

    companion object {
        private val log = LoggerFactory.getLogger(AuthPortAdapter::class.java)
        private const val DEFAULT_ROLE = "USER"
        private const val REFRESH_EXPIRES_IN = 604800L // 7 days (not implemented yet)
    }
}
